components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/LegacyJobOrchestrator.cs

```diff
@@ -249,113 +246,134 @@ public class LegacyJobOrchestrator : IJobOrchestrator, IPublisherConfigServices,
+        /// Deserialize string representing published nodes file into PublishedNodesEntryModel entries and
+        /// run schema validation if that is enables.
+        /// </summary>
+        /// <param name="content"></param>
+        /// <returns></returns>
+        private IEnumerable<PublishedNodesEntryModel> DeserializePublishedNodes(string content) {
+            if (!File.Exists(_legacyCliModel.PublishedNodesSchemaFile)) {
+                _logger.Information("Validation schema file {PublishedNodesSchemaFile} does not exist or is disabled, ignoring validation of {publishedNodesFile} file...",
+                _legacyCliModel.PublishedNodesSchemaFile, _legacyCliModel.PublishedNodesFile);
+                return _publishedNodesJobConverter.Read(content, null);
+            }
+            else {
+                using (var fileSchemaReader = new StreamReader(_legacyCliModel.PublishedNodesSchemaFile)) {
+                    return _publishedNodesJobConverter.Read(content, fileSchemaReader);
+                }
+            }
+        }
+
+        private bool IsSameDataSet(PublishedNodesEntryModel entry1, PublishedNodesEntryModel entry2) {
+            return entry1.HasSameGroup(entry2)
+                && entry1.DataSetWriterId == entry2.DataSetWriterId
+                && entry1.DataSetPublishingInterval == entry2.DataSetPublishingInterval;
+        }
+
+        private void RefreshJobs(IEnumerable<PublishedNodesEntryModel> entries) {
+            var availableJobs = new Dictionary<string, JobProcessingInstructionModel>();
+            var assignedJobs = new Dictionary<string, JobProcessingInstructionModel>();
+
+            _lockJobs.Wait();
+            try {
+                // Create JobId to WorkerId dictionary for fast lookup.
+                var jobIdDict = _assignedJobs
+                    .Select(kvp => Tuple.Create(kvp.Value.Job.Id, kvp.Key))
+                    .ToDictionary(tuple => tuple.Item1);
+
+                var jobs = _publishedNodesJobConverter.ToWriterGroupJobs(entries, _legacyCliModel);
+                if (jobs.Any()) {
+                    foreach (var job in jobs) {
+                        var newJob = ToJobProcessingInstructionModel(job);
+                        if (string.IsNullOrEmpty(newJob?.Job?.Id)) {
+                            continue;
+                        }
+
+                        var newJobId = newJob.Job.Id;
+                        if (jobIdDict.ContainsKey(newJobId)) {
+                            assignedJobs[jobIdDict[newJobId].Item2] = newJob;
+                        }
+                        else {
+                            availableJobs.Add(newJobId, newJob);
+                        }
+                    }
+                }
+
+                // Update local state.
+                _availableJobs = availableJobs;
+                _assignedJobs = assignedJobs;
+
+                AdjustMaxWorkersAgentConfig();
+
+                _publishedNodesEntries.Clear();
```

Comment:
this needs to be guarded by _lockConfig, ideally outside of the _lockJobs. Alternatively taken out of this method.

Reply:
Took updates of `_publishedNodesEntries` out of `RefreshJobs()` method.